torch/fx/passes/utils/matcher_utils.py

```diff
@@ -380,3 +387,41 @@ def backtracking(anchor_index, match):
         logger.info("Matches returned: %s", matches)
 
         return matches
+
+def _split_to_graph_and_name_nodes_map(
+    gm: GraphModule,
+) -> Tuple[GraphModule, Dict[str, Node]]:
+    from torch.utils._pytree import tree_flatten, tree_unflatten
+    from torch.fx.graph import _PyTreeInfo
+    name_nodes_map = {}
+    for n in gm.graph.nodes:
+        if n.op == "output":
+            output = tree_unflatten(n.args[0], gm._out_spec)
+            assert isinstance(output, tuple), "Expecting the pattern graph to return a tuple"
+            assert len(output) >= 2, "Expecting the pattern graph to have at least two outputs"
+            *out, name_nodes_map = output
+            flattened, out_spec = tree_flatten(out)
+            assert isinstance(name_nodes_map, Dict), "Expecting the input graph to have a dict output as the last element"
+            n.args = (flattened,)
+            orig_pytree_info = gm._graph._codegen.pytree_info
+            gm._graph._codegen.pytree_info = _PyTreeInfo(orig_pytree_info.orig_args, orig_pytree_info.in_spec, out_spec)
+    gm.recompile()
+    return gm, name_nodes_map
+
+
+class SubgraphMatcherWithNameNodesMap(SubgraphMatcher):
```

Comment:
Need documentation on this. Especially 

- Pattern graph's output convention about `{"conv": conv, "relu": relu}`
- match()'s return InternalMatch. name_nodes_map now has a different meaning.